File: src/android/com/bluvision/beacons/plugin/BeaconInteractor.java
package com.bluvision.beacons.plugin;

import com.bluvision.sdk.beacons.Beacon;
import com.bluvision.sdk.beacons.BeaconManager;
import com.bluvision.sdk.beacons.Blufi;
import com.bluvision.sdk.beacons.SBeacon;
import com.bluvision.sdk.cloud.Device;
import com.bluvision.sdk.cloud.Error;
import com.bluvision.sdk.cloud.Location;
import com.bluvision.sdk.cloud.Provisioner;
import com.bluvision.sdk.cloud.Template;
import com.bluvision.sdk.cloud.User;
import com.bluvision.sdk.cloud.callbacks.LocationListCallback;
import com.bluvision.sdk.cloud.callbacks.ProvisioningCallback;
import com.bluvision.sdk.cloud.callbacks.UserCallback;
import com.bluvision.sdk.constants.Distance;
import com.bluvision.sdk.utilities.ErrorCode;

import android.app.Activity;
import android.content.Context;
import android.util.Log;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class BeaconInteractor implements BeaconManager.BeaconListener {
    private BeaconManager mBeaconManager;
    private Map<String, Beacon> beaconList = new HashMap<>();
    private Map<String, Template> templateList = new HashMap<>();
    private BeaconInteractorCallback mBeaconCallback;
    private boolean scanning;
    private Context context;
    private Activity activity;

    public void init(Context context, BeaconInteractorCallback beaconCallback, Activity activity) {
        this.context = context;
        this.activity = activity;
        mBeaconCallback = beaconCallback;
    }

    public void startScan() {
        if(mBeaconManager == null) {
            mBeaconManager = BluVisionManagerHandler.getInstance(context, this);
        }
        if (mBeaconManager != null && !scanning) {
            mBeaconManager.startScanning();
            scanning = true;
        }
    }

    public void stopScan() {
        if (mBeaconManager != null && scanning) {
            mBeaconManager.stopScanning();
            scanning = false;
            BluVisionManagerHandler.destroyManager();
            mBeaconManager = null;
        }
    }

    @Override
    public void onFound(final Beacon beacon) {
        if (beacon instanceof SBeacon) {
            beaconList.put(((SBeacon) beacon).getIdentifier().toString(), beacon);
            if (mBeaconCallback != null) {
                mBeaconCallback.beaconFound(beacon);
            }
        }


    }

    @Override
    public void onLost(final Beacon beacon) {
        if (beacon instanceof SBeacon) {
            beaconList.remove(((SBeacon) beacon).getIdentifier().toString());
            if (mBeaconCallback != null) {
                mBeaconCallback.beaconLost(beacon);
            }
        }
    }

    @Override
    public void onChange(Beacon beacon, Distance distance) {

    }

    @Override
    public void onScanningFail(int errorCode) {
        if (errorCode == ErrorCode.BEACON_SCANNER_BLUETOOTH_NOT_ENABLED) {
            mBeaconCallback.enableBluetooth();
        }

        if (errorCode == ErrorCode.BEACON_SCANNER_INSUFFICIENT_PERMISSIONS) {
            mBeaconCallback.requestLocationPermissions();
        }
    }

    public void signIn(String token) {
        User.AuthenticateWithAPIToken(token, context, new UserCallback() {
            @Override
            public void onComplete(User user, Error error) {
                Log.i("authenticate", "User " + user + "error " + error);
                if (user != null) {
                    mBeaconCallback.signInSuccess(user);
                    startScan();
                } else {
                    mBeaconCallback.signInFailure(error);
                    Log.i("authenticate", error.getMessage());
                }
            }
        });
    }

    public void signOut() {
        User user = User.getCurrentUser();
        if(user != null){
            user.signOut();
            beaconList.clear();
        }
    }

    public void loadTemplates(String id) {
        SBeacon beacon = (SBeacon) beaconList.get(id);
        templateList = new HashMap<>();
        if (beacon != null) {
            if (beacon instanceof Blufi) {
                Template.getBlufiTemplates(callback);
            } else {
                Template.getBeaconTemplates(beacon.getIdentifier().toString(), callback);
            }
        }
    }

    private Template.TemplateListCallback callback = new Template.TemplateListCallback() {
        @Override
        public void onComplete(List<? extends Template> list, Error error) {
            if (error == null) {
                //There are no templates for this device, please create one to continue
                mBeaconCallback.loadTemplateSucess(list);
                for (Template template : list) {
                    templateList.put("" + template.getTemplateId(), template);
                }
            } else {
                if (error.getCode() > 50000) {
                    mBeaconCallback.loadTemplateError("This device cannot be provisioned. Please contact support for assistance.");

                    if (error.getCode() == 50404) {
                        mBeaconCallback.loadTemplateError("Device already Provisioned. This Device is already provisioned for use with " +
                                                          "another project.");
                    }
                } else {
                    mBeaconCallback.loadTemplateError(error.getMessage());
                    Log.e("templates", error.getMessage());
                }
            }
        }
    };

    private void getLocations(final String beacon, final int template, final String beaconName, final String notes) {
        User.getCurrentUser().getCurrentProject().getLocations(new LocationListCallback() {
            @Override
            public void onComplete(List<Location> list, Error error) {
                if (list == null || list.size() == 0) {
                    Log.e("error", "for this sample we need a location");
                    return;
                }
                provision(list.get(0), beacon, template, beaconName, notes);
            }
        });

    }

    private void provision(Location location, String beacon, int template, String beaconName, String notes) {
        SBeacon mDevice = (SBeacon) beaconList.get(beacon);
        Provisioner mProvisioner = new Provisioner(mDevice, template, activity);

        mProvisioner.setLocationIdentifier(location.getLocationId());
        mProvisioner.setLatitude(25.8); // Get this from a map marker
        mProvisioner.setLongitude(80.2);
        mProvisioner.setDeviceName(beaconName);
        mProvisioner.setNotes(notes);

        mProvisioner.startProvisioning(new ProvisioningCallback() {
            @Override
            public void onSuccess(Device device) {
                mBeaconCallback.provisionSuccess();
            }

            @Override
            public void onError(Error error) {
                mBeaconCallback.provisionError(error.getMessage());
            }

            @Override
            public void onStatusChange(String s) {
                mBeaconCallback.provisionStatusChange(s);
            }

            @Override
            public void onDescriptionChange(String s) {
                mBeaconCallback.provisiondescriptionChange(s);
            }
        });
    }

    public void provisionBeacon(String beacon, int template, String beaconName, String notes) {
        getLocations(beacon, template, beaconName, notes);
    }
}

File: src/ios/com/bluvision/beacons/plugin/BeaconInteractor.m
//
//  BeaconInteractor.m
//  BluPro
//

#import "BeaconInteractor.h"

#import <Bluvision/BLUBeaconManager+Private.h>

@interface BZCClient ()

- (void)enableNetworkLogging;

@end

@interface BeaconInteractor()<BLUBeaconManagerDelegate, BZCProvisionerDelegate>

@property (nonatomic, strong) BZCProvisioner *provisioner;
@property (nonatomic, strong) NSMutableDictionary *beaconList;
@property (nonatomic, strong) NSMutableDictionary *templateList;
@property (nonatomic, strong) BLUBeaconManager *beaconManager;
@property (nonatomic, strong) BZCDevice *provisionedDevice;
@property (nonatomic, weak) id<BeaconInteractorDelegate> delegate;

@end

@implementation BeaconInteractor

BOOL scanning = false;

- (id)initWithDelegate:(id <BeaconInteractorDelegate>)delegate {

    if (self = [super init])
    {
        self.delegate = delegate;
        self.beaconList = [[NSMutableDictionary alloc] init];
        self.beaconManager = [[BLUBeaconManager alloc] initWithDelegate:self];
    }
    return self;
}

- (void)signIn:(NSString *)token device:(NSString *)deviceType {

    scanning = false;
    [[BZCClient client] enableNetworkLogging];

    [BZCUser authenticateWithAPIToken:token completion:^(BZCUser * _Nullable user, NSError * _Nullable error) {
        if (error == nil && user != nil) {
            [self.delegate signInSuccess:user];
            [self startScan:deviceType];
        } else {
            [self.delegate signInFailure:error];
        }
    }];
}

- (void)startScan:(NSString *)deviceType {

    if (self.beaconManager != nil && scanning == false) {
        scanning = true;
        if ([deviceType isEqualToString:@"SBeacon"]) {
            [self.beaconManager startScanningForBeacons];
        } else {
            [self.beaconManager startScanningForBluFis];
        }
    }
}

- (void)stopScan {

    if (self.beaconManager != nil && scanning == true) {
        [self.beaconManager stopScanning];
        scanning = false;
    }
}

- (void)loadTemplates:(NSString *)beacondentifier {

    if (self.provisioner != nil) {
        self.provisioner = nil;
    }

    void (^completion)(NSArray *templates, NSError *error) = ^(NSArray *templates, NSError *error) {
        if (error) {
            if (error.code >= 50000) {
                if (error.code == 50404) {
                    NSString *error = @"This transmitter is already provisioned for use with another project";
                    [self.delegate loadTemplateError:error];
                } else {
                    NSString *error = @"This device cannot be provisioned, Please contact support for assistance";
                    [self.delegate loadTemplateError:error];
                }
            }
            else {
                [self.delegate loadTemplateError:error.localizedDescription];
            }
        } else {
            for (BZCTemplate *template in templates) {
                [self.templateList setObject:template forKey:template.identifier.stringValue];
            }
            [self.delegate loadTemplateSucess:templates];
        }
    };

    BLUSBeacon *configurableBeacon = [self.beaconList objectForKey:beacondentifier];
    self.provisioner = [[BZCProvisioner alloc] initWithConfigurableBeacon:configurableBeacon];

    switch (deviceTypeFromString(self.provisioner.deviceType)) {
        case BZCDeviceTypeBeacon:
            [BZCTemplate beaconTemplatesForDeviceIdentifier:self.provisioner.configurableBeacon.identifier completion:completion];
            break;
        case BZCDeviceTypeBluFi:
            [BZCTemplate blufiTemplatesWithCompletion:completion];
            break;
        default:
            break;
    }
}

- (void)provisionBeaconForTemplate:(int)templateId beaconName:(NSString *)name notes:(NSString *)notes {

    if (self.provisioner != nil) {

        self.provisioner.templateIdentifier = [NSNumber numberWithInt:templateId];
        self.provisioner.deviceFriendlyName = name;
        self.provisioner.notes = notes;
        [self getLocations];
    }
}

- (void)getLocations {

    [[[BZCUser currentUser] currentProject] locationsWithCompletion:^(NSArray<BZCLocation *> * _Nullable locations, NSError * _Nullable error) {
        if (locations.count == 0) {
            NSLog(@"Locations are needed for this example, locations can be auto generated if there are none but require a geo location that is not (0,0). "
                  "Please create a location on your cloud or supply a geo location to the provisioner object in prepareForSegue:");
        }

        if (locations != nil && locations.count > 0) {
            BZCLocation *location = locations[0];
            self.provisioner.location = location;
            self.provisioner.geolocation = [[CLLocation alloc] initWithCoordinate:location.overlayCenterCoordinate altitude:0 horizontalAccuracy:0 verticalAccuracy:0 timestamp:[NSDate date]];
            [self provisionBeacon];
        }
    }];
}

- (void)provisionBeacon {

    self.provisioner.delegate = self;

    [self.provisioner startProvisioningWithCompletion:^(BZCDevice * _Nullable device, NSError * _Nullable error) {

        if (error) {

            NSLog(@"%@", error.localizedDescription);
            [self.delegate provisionError:error.localizedDescription];
            return;
        }
        self.provisionedDevice = device;
        [self.delegate provisionSuccess];
    } recoverableConnectionFailureBlock:^BOOL(NSError * _Nullable error) {
        if (error.code == 5022) {

            [self.delegate provisionError:error.localizedDescription];
        }
        return NO;
    }];

}

#pragma mark BLUBeaconManagerDelegate

- (void)beaconManager:(BLUBeaconManager *)manager didFindBeacon:(BLUBeacon *)beacon {

    if (beacon != nil && [beacon isKindOfClass:[BLUSBeacon class]]) {

        BLUSBeacon *device = (BLUSBeacon *)beacon;
        [self.beaconList setObject:device forKey:device.identifier.stringValue];
        [self.delegate beaconFound:device];
    }
}

- (void)beaconManager:(BLUBeaconManager *)manager didLoseBeacon:(BLUBeacon *)beacon {

    if (beacon != nil && [beacon isKindOfClass:[BLUSBeacon class]]) {

        BLUSBeacon *device = (BLUSBeacon *)beacon;
        [self.beaconList removeObjectForKey:device.identifier.stringValue];
        [self.delegate beaconLost:device];
    }
}

- (void)beaconManager:(BLUBeaconManager *)manager didFailWithError:(nullable NSError *)error {

    NSLog(@"Error = %@", error.localizedDescription);
}

#pragma mark - provisioner delegate

- (void)provisioner:(BZCProvisioner *)provisioner didChangeStatus:(NSString *)status {

    [self.delegate provisionStatusChange:status];
}

- (void)provisioner:(BZCProvisioner *)provisioner didChangeDescription:(NSString *)description {

    [self.delegate provisionDescriptionChange:description];
}

- (void)provisioner:(nonnull BZCProvisioner *)provisioner willRetryProvisioningWithError:(nonnull NSError *)error {

    NSLog(@"%@", error.localizedDescription);
}


@end
